### PR TITLE
MLFlow and AzureML for Libero Fine-Tuning

### PR DESCRIPTION
**Pull request by noraabiakar**
## Summary
This PR introduces a full AzureML pipeline for fine-tuning on the Libero dataset, including all necessary configuration files, documentation, and code changes to support MLflow tracking.

## Key changes
- New folder `finetuning_with_azureml` in the `examples` directory to reproduce the training in Azure ML, including a README that describes how to:
    - Set up required Azure resources (workspace, compute instances, storage, and datastores)
    - Create and version AzureML environments with Docker support
    - Download the Libero dataset and upload it to Azure storage
    - Generate and store normalization statistics for the dataset
    - Run fine-tuning jobs on AzureML with configurable parameters, checkpointing, and resume/overwrite options

- Added YAML job templates and Dockerfile for reproducible environment and job setup in AzureML.
- Modified `script/train.py` to support MLFlow in addition to WandB for tracking metrics.

### CODE REVIEW

File: examples/finetune_with_azureml/train.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ --num-train-steps ${{inputs.num_train_steps}}
+ --num-workers 8 
+ --fsdp-devices 1 
+ --batch-size 32 
+ --mlflow-enabled
+ --no-wandb-enabled
+ ${{inputs.extra_flags}}
+environment: azureml:<env_name>:<env_version>
+inputs:
+  pi0_experiment_name: <pi0_exp_name>
```
Comment (fbeltrao) [inline examples/finetune_with_azureml/train.yaml:16]: Would it make sense to take the value of `pi0_experiment_name` from L36 `experiment_name: p0_libero_finetune`?

File: examples/finetune_with_azureml/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ENV UV_PROJECT_ENVIRONMENT=/.venv
+
+COPY . .
+
+# Install the project's dependencies using the lockfile and settings
+RUN uv venv --python 3.11.9 $UV_PROJECT_ENVIRONMENT
+RUN --mount=type=cache,target=/root/.cache/uv \
+    GIT_LFS_SKIP_SMUDGE=1 uv sync --frozen --no-dev \
+    && uv pip install azureml-mlflow \
+    && uv pip install orbax-checkpoint==0.11.1
```
Comment (fbeltrao) [inline examples/finetune_with_azureml/Dockerfile]: Shall we add a comment on the downgrade of orbax-checkpoint?

**Review by fbeltrao — APPROVED** [review]

File: scripts/train.py [hunk 3]
```diff
@@ -46,11 +48,24 @@ def format(self, record):
...
         wandb.init(mode="disabled")
-        return
 
+    # Initialize MLflow and yield the context manager if mlflow is enabled. Otherwise, yield a null context.
+    if config.mlflow_enabled:
+        with mlflow.start_run() as run:
+            init_mlflow(config)
+            yield run
+    else:
+        yield nullcontext()
```
Comment (fbeltrao) [inline scripts/train.py:66]: Could it be that the `Run` instance returned by `wandb.init` is a [context manager](https://github.com/wandb/wandb/blob/v0.20.1/wandb/sdk/wandb_run.py#L3664). If that is true, should we return it instead of a null context?

File: scripts/train.py [hunk 4]
```diff
@@ -69,6 +84,30 @@ def init_wandb(config: _config.TrainConfig, *, resuming: bool, log_code: bool =
...
+    mlflow.log_param("resume_from_checkpoint", config.resume)
+    mlflow.log_param("overwrite_checkpoint_dir", config.overwrite)
+
+
+def log_metrics(config: _config.TrainConfig, metrics: dict[str, Any], step: int):
+    if config.wandb_enabled:
+        wandb.log(metrics, step=step)
+    if config.mlflow_enabled:
+        for k, v in metrics.items():
+            mlflow.log_metric(k, float(v), step=step)
```
Comment (fbeltrao) [inline scripts/train.py]: nit: seems like `mlflow.log_metrics` does a single roundtrip to server ([via log_batch](https://github.com/mlflow/mlflow/blob/7d94f033e223ebb00d781f38fc685a0cbdfa2ba4/mlflow/tracking/fluent.py#L1117)), while each call to `mlflow.log_metric` will be a [single roundtrip](https://github.com/mlflow/mlflow/blob/7d94f033e223ebb00d781f38fc685a0cbdfa2ba4/mlflow/tracking/fluent.py#L981)

**Review by fbeltrao — APPROVED** [review]
Great job @noraabiakar !

File: examples/finetune_with_azureml/train.yaml (hunk 1 above)
Comment (noraabiakar) [reply examples/finetune_with_azureml/train.yaml:16]: The "pi0_experiment_name" is ultimately used to name the subfolder where the checkpoints are used. If we always use the same experiment name ("p0_libero_finetune" from L37) we would need to overwrite the checkpoints or resume from the last one every time. I see 2 options:
1. Rename "pi0_experiment_name" to "checkpoint_subdir"
2. Make "experiment_name" configurable and use that on L4 and L37. (This would change the AzureML experiment which groups jobs). 
What do you think?

File: scripts/train.py (hunk 3 above)
Comment (noraabiakar) [reply scripts/train.py:66]: It can return a context. But what if both wandb and mlflow are enabled? Then we'd have to use nested contexts. I thought this implementation was simpler. I see 3 options: 
1. We use nested contexts (if you have an idea of how to do this in an elegant way maybe you can show me)
2. We only allow one of mlflow or wandb and use always a single context
3. We keep this implementation

Comment (fbeltrao) [reply scripts/train.py:66]: I'd suggest 1 or 2.
